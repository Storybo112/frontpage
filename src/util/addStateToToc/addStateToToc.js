const githubDocsBaseUrl = 'https://github.com/storybookjs/storybook/tree/next';

function prefixPath(path, prefix) {
  return path ? `${prefix}/${path}` : prefix;
}

module.exports = function addStateToToc(items, pathPrefix = '/docs') {
  return items.map((item) => {
    const isIndexPage =
      (item.type === 'heading' && !item.redirectPath) ||
      (item.type.match(/link|sub-page/) && item.subPages);

    const path = prefixPath(item.pathSegment, pathPrefix);

    return {
      ...item,
      ...(item.type.match(/link|heading|sub-page/) && {
        path: item.redirectPath ? prefixPath(item.redirectPath, pathPrefix) : path,
        githubUrl: `${githubDocsBaseUrl}${path}${isIndexPage ? '/index' : ''}.md`,
      }),
      ...(item.children && { children: addStateToToc(item.children, path) }),
      ...(item.subPages && { subPages: addStateToToc(item.subPages, path) }),
    };
  });
};
