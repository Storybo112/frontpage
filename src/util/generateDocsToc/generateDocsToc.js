const fs = require('fs');
const path = require('path');
const matter = require('gray-matter');

const NON_CONTENT_DIRECTORIES = ['assets', 'snippets', 'versions'];

function getMetadata(filePath) {
  const fileContents = fs.readFileSync(filePath, 'utf8');
  const {
    data: { navTitle, title, ...data },
  } = matter(fileContents);
  return {
    title: navTitle || title,
    ...data,
  };
}

/**
 * This is a factory function only for dependency injection.
 *
 * We need to sort both files (with a `navOrder` in their metadata) and directories (with a
 * `navOrder` in their `index.md`'s metadata).
 */
function makeSortFilesAndDirectories(generateDocsToc, pathToFiles, docsRoot) {
  return function sortFilesAndDirectories(a, b) {
    const aFilePath = path.join(pathToFiles, a);
    const bFilePath = path.join(pathToFiles, b);

    const aIsDirectory = fs.lstatSync(aFilePath).isDirectory();
    const bIsDirectory = fs.lstatSync(bFilePath).isDirectory();

    let aOrder = 0;
    let bOrder = 0;

    if (aIsDirectory) {
      const childItems = generateDocsToc(aFilePath, docsRoot);
      const indexItem = childItems.find((child) => child.pathSegment === 'index');
      if (indexItem) {
        aOrder = indexItem.navOrder;
      }
    } else {
      const metaData = getMetadata(aFilePath);
      aOrder = metaData.navOrder;
    }

    if (bIsDirectory) {
      const childItems = generateDocsToc(bFilePath, docsRoot);
      const indexItem = childItems.find((child) => child.pathSegment === 'index');
      if (indexItem) {
        bOrder = indexItem.navOrder;
      }
    } else {
      const metaData = getMetadata(bFilePath);
      bOrder = metaData.navOrder;
    }

    return aOrder - bOrder;
  };
}

/**
 * @param pathToDocs string - Path to docs directory, relative to project root
 *
 * @param docsRoot string - Private; Path to docs root, relative to project root (used for recursion)
 *
 * @returns TOC array
 */
module.exports = function generateDocsToc(
  pathToFiles = 'src/content/docs',
  docsRoot = pathToFiles
) {
  const files = fs.readdirSync(pathToFiles);

  if (files.includes('toc.js')) {
    // eslint-disable-next-line global-require, import/no-dynamic-require
    const { toc } = require(path.join('../../../', pathToFiles, 'toc.js'));
    return toc;
  }

  const toc = [];
  files
    .filter((file) => !NON_CONTENT_DIRECTORIES.includes(file))
    .sort(makeSortFilesAndDirectories(generateDocsToc, pathToFiles, docsRoot))
    .forEach((file) => {
      const filePath = path.join(pathToFiles, file);
      const isDirectory = fs.lstatSync(filePath).isDirectory();

      if (isDirectory) {
        const dirPathSegment = file.replace(/^\d+-/, '');
        const childItems = generateDocsToc(filePath, docsRoot);

        if (childItems) {
          let title;
          let redirectPath;

          /**
           * We have to do the sub-page processing before the child items processing, because we
           * need the index child that is removed in the child items processing.
           */
          const subPageItems = childItems.filter((child) => {
            return child.type === 'sub-page';
          });
          const hasSubPages = subPageItems.length > 0;

          let subPages;
          let subPagePathSegments;
          if (hasSubPages) {
            subPagePathSegments = [
              // Include implicit index sub-page
              'index',
              ...subPageItems.map((subPage) => subPage.pathSegment),
            ];
            subPages = subPageItems.map(
              ({ activeSubPage, pathSegment, subPages: nestedSubPages, type }) => ({
                activeSubPage,
                pathSegment,
                subPagePathSegments,
                ...(nestedSubPages
                  ? {
                      subPages: nestedSubPages
                        // We don't want the index sub-page to be a sub-page of itself
                        .filter((nestedSubPage) => nestedSubPage.pathSegment !== 'index')
                        // We want the `activeSubPage` and `subPagePathSegments` of the _parent_ sub-page, so we overwrite them here
                        .map((nestedSubPage) => ({
                          ...nestedSubPage,
                          activeSubPage,
                          subPagePathSegments,
                        })),
                      activeSubPage,
                    }
                  : {}),
                type,
              })
            );
          }

          const indexItemIndex = childItems.findIndex((child) => child.pathSegment === 'index');

          const hasNestedSubPages =
            indexItemIndex !== -1 && childItems[indexItemIndex].type === 'sub-page';

          if (indexItemIndex !== -1) {
            const metaData = getMetadata(path.join(filePath, childItems[indexItemIndex].file));
            title = metaData.title;

            if (metaData.redirectTo) {
              redirectPath = path.join(dirPathSegment, metaData.redirectTo);
            }

            // Remove the index item, as it is represented by the parent item
            childItems.splice(indexItemIndex, 1);
          }

          const children = childItems
            /**
             * Items always have a `navOrder` and, when in a group, a `navGroup.order`.
             * This sorts by `navGroup.order` (falling back to `navOrder`) first, then `navOrder`.
             */
            .sort(
              (a, b) =>
                (a.navGroup?.order || a.navOrder) - (b.navGroup?.order || b.navOrder) ||
                a.navOrder - b.navOrder
            )
            .reduce(
              (
                acc,
                {
                  navGroup,
                  // Remove these from the child items, as they're not needed beyond this point
                  file: ignoreFile,
                  navOrder: ignoreNavOrder,
                  ...child
                }
              ) => {
                if (navGroup) {
                  const group = acc.find((c) => c.title === navGroup.title);
                  if (group) {
                    group.children.push(child);
                  } else {
                    acc.push({
                      pathSegment: '',
                      title: navGroup.title,
                      type: 'menu',
                      children: [child],
                    });
                  }
                } else {
                  acc.push(child);
                }
                return acc;
              },
              []
            );

          // Push the non-page item, with all of its processed children/sub-pages
          toc.push({
            pathSegment: dirPathSegment,
            ...(redirectPath ? { redirectPath } : {}),
            ...(title ? { title } : {}),
            ...(hasSubPages
              ? {
                  subPages,
                  subPagePathSegments,
                  ...(hasNestedSubPages
                    ? { type: 'sub-page', activeSubPage: dirPathSegment }
                    : { type: 'link', activeSubPage: 'index' }),
                }
              : { type: 'heading', children }),
          });
        }
      } else if (file.endsWith('.md')) {
        const { isSubPage, ...metaData } = getMetadata(filePath);
        const filePathSegment = file.replace(/^(?:\d+-)?(.*)\.md$/, '$1');

        // Push the page item
        toc.push({
          file,
          pathSegment: filePathSegment,
          ...(isSubPage ? { type: 'sub-page', activeSubPage: filePathSegment } : { type: 'link' }),
          ...metaData,
        });
      }
    });

  return toc;
};
