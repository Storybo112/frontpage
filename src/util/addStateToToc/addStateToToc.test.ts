import path from 'path';

import generateDocsToc from '../generateDocsToc/generateDocsToc';
import addStateToToc from './addStateToToc';

const getToc = (mockPath) =>
  generateDocsToc(path.join('src/util/generateDocsToc/__mocks__', mockPath));

describe('addStateToToc', () => {
  it('adds state to basic TOC', () => {
    const result = addStateToToc(getToc('basic'));
    expect(result).toMatchSnapshot();
  });

  it('handles heading index items', () => {
    const result = addStateToToc(getToc('basic'));
    const headingItem = result.find((item) => item.type === 'heading');
    expect(headingItem).toHaveProperty(
      'githubUrl',
      'https://github.com/storybookjs/storybook/tree/next/docs/writing-stories/index.md'
    );
  });

  it('handles redirects', () => {
    const result = addStateToToc(getToc('with-redirect'));
    const redirectedItem = result.find((item) => item.redirectPath);
    expect(redirectedItem).toHaveProperty('path', '/docs/parent/target');
  });

  it('handles sub-pages', () => {
    const result = addStateToToc(getToc('with-sub-pages'));
    expect(result).toMatchSnapshot();
  });

  it('handles nested sub-pages', () => {
    const result = addStateToToc(getToc('with-nested-sub-pages'));
    expect(result).toMatchSnapshot();
  });
});
