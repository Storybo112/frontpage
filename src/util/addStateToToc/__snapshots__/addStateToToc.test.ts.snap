// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addStateToToc adds state to basic TOC 1`] = `
[
  {
    "children": [
      {
        "githubUrl": "https://github.com/storybookjs/storybook/tree/next/docs/writing-stories/parameters.md",
        "path": "/docs/writing-stories/parameters",
        "pathSegment": "parameters",
        "title": "Parameters (should be first)",
        "type": "link",
      },
      {
        "githubUrl": "https://github.com/storybookjs/storybook/tree/next/docs/writing-stories/decorators.md",
        "path": "/docs/writing-stories/decorators",
        "pathSegment": "decorators",
        "title": "Decorators (should be second)",
        "type": "link",
      },
    ],
    "githubUrl": "https://github.com/storybookjs/storybook/tree/next/docs/writing-stories/index.md",
    "path": "/docs/writing-stories",
    "pathSegment": "writing-stories",
    "title": "How to write stories (should be first)",
    "type": "heading",
  },
  {
    "children": [
      {
        "githubUrl": "https://github.com/storybookjs/storybook/tree/next/docs/writing-docs/autodocs.md",
        "path": "/docs/writing-docs/autodocs",
        "pathSegment": "autodocs",
        "title": "Automatic documentation and Storybook (should be first)",
        "type": "link",
      },
      {
        "githubUrl": "https://github.com/storybookjs/storybook/tree/next/docs/writing-docs/mdx.md",
        "path": "/docs/writing-docs/mdx",
        "pathSegment": "mdx",
        "title": "MDX (should be second)",
        "type": "link",
      },
    ],
    "githubUrl": "https://github.com/storybookjs/storybook/tree/next/docs/writing-docs/index.md",
    "path": "/docs/writing-docs",
    "pathSegment": "writing-docs",
    "title": "How to document components (should be second)",
    "type": "heading",
  },
  {
    "file": "12-faq.md",
    "githubUrl": "https://github.com/storybookjs/storybook/tree/next/docs/faq.md",
    "path": "/docs/faq",
    "pathSegment": "faq",
    "title": "Frequently Asked Questions (should be third)",
    "type": "link",
  },
]
`;

exports[`addStateToToc handles nested sub-pages 1`] = `
[
  {
    "children": [
      {
        "githubUrl": "https://github.com/storybookjs/storybook/tree/next/docs/parent/before.md",
        "path": "/docs/parent/before",
        "pathSegment": "before",
        "title": "Before (should be first)",
        "type": "link",
      },
      {
        "activeSubPage": "index",
        "githubUrl": "https://github.com/storybookjs/storybook/tree/next/docs/parent/test/index.md",
        "path": "/docs/parent/test",
        "pathSegment": "test",
        "subPagePathSegments": [
          "index",
          "api",
          "migration",
        ],
        "subPages": [
          {
            "activeSubPage": "api",
            "githubUrl": "https://github.com/storybookjs/storybook/tree/next/docs/parent/test/api/index.md",
            "path": "/docs/parent/test/api",
            "pathSegment": "api",
            "subPagePathSegments": [
              "index",
              "api",
              "migration",
            ],
            "subPages": [
              {
                "activeSubPage": "api",
                "githubUrl": "https://github.com/storybookjs/storybook/tree/next/docs/parent/test/api/nested-one.md",
                "path": "/docs/parent/test/api/nested-one",
                "pathSegment": "nested-one",
                "subPagePathSegments": [
                  "index",
                  "api",
                  "migration",
                ],
                "type": "sub-page",
              },
              {
                "activeSubPage": "api",
                "githubUrl": "https://github.com/storybookjs/storybook/tree/next/docs/parent/test/api/nested-two.md",
                "path": "/docs/parent/test/api/nested-two",
                "pathSegment": "nested-two",
                "subPagePathSegments": [
                  "index",
                  "api",
                  "migration",
                ],
                "type": "sub-page",
              },
            ],
            "type": "sub-page",
          },
          {
            "activeSubPage": "migration",
            "githubUrl": "https://github.com/storybookjs/storybook/tree/next/docs/parent/test/migration.md",
            "path": "/docs/parent/test/migration",
            "pathSegment": "migration",
            "subPagePathSegments": [
              "index",
              "api",
              "migration",
            ],
            "type": "sub-page",
          },
        ],
        "title": "Test",
        "type": "link",
      },
      {
        "githubUrl": "https://github.com/storybookjs/storybook/tree/next/docs/parent/after.md",
        "path": "/docs/parent/after",
        "pathSegment": "after",
        "title": "After (should be third)",
        "type": "link",
      },
    ],
    "githubUrl": "https://github.com/storybookjs/storybook/tree/next/docs/parent/index.md",
    "path": "/docs/parent",
    "pathSegment": "parent",
    "title": "Parent",
    "type": "heading",
  },
]
`;

exports[`addStateToToc handles sub-pages 1`] = `
[
  {
    "children": [
      {
        "githubUrl": "https://github.com/storybookjs/storybook/tree/next/docs/parent/before.md",
        "path": "/docs/parent/before",
        "pathSegment": "before",
        "title": "Before (should be first)",
        "type": "link",
      },
      {
        "activeSubPage": "index",
        "githubUrl": "https://github.com/storybookjs/storybook/tree/next/docs/parent/test/index.md",
        "path": "/docs/parent/test",
        "pathSegment": "test",
        "subPagePathSegments": [
          "index",
          "api",
          "migration",
        ],
        "subPages": [
          {
            "activeSubPage": "api",
            "githubUrl": "https://github.com/storybookjs/storybook/tree/next/docs/parent/test/api.md",
            "path": "/docs/parent/test/api",
            "pathSegment": "api",
            "subPagePathSegments": [
              "index",
              "api",
              "migration",
            ],
            "type": "sub-page",
          },
          {
            "activeSubPage": "migration",
            "githubUrl": "https://github.com/storybookjs/storybook/tree/next/docs/parent/test/migration.md",
            "path": "/docs/parent/test/migration",
            "pathSegment": "migration",
            "subPagePathSegments": [
              "index",
              "api",
              "migration",
            ],
            "type": "sub-page",
          },
        ],
        "title": "Test",
        "type": "link",
      },
      {
        "githubUrl": "https://github.com/storybookjs/storybook/tree/next/docs/parent/after.md",
        "path": "/docs/parent/after",
        "pathSegment": "after",
        "title": "After (should be third)",
        "type": "link",
      },
    ],
    "githubUrl": "https://github.com/storybookjs/storybook/tree/next/docs/parent/index.md",
    "path": "/docs/parent",
    "pathSegment": "parent",
    "title": "Parent",
    "type": "heading",
  },
]
`;
